DRA: automated upgrade/downgrade testing

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Promotion of any feature, in this case core DRA to GA, depends on (so far) manually testing the upgrade/downgrade path. It's hard to document what exactly was tested in a way that others can verify the procedure and/or replicate it.

This PR adds helper packages for upgrading/downgrading with local-up-cluster.sh and uses that to test some DRA scenarios. Those tests run as part of a special test/e2e_dra test.

#### Which issue(s) this PR is related to:

https://github.com/kubernetes/kubernetes/issues/128965
KEP: https://github.com/kubernetes/enhancements/issues/4381

#### Special notes for your reviewer:

How to implement the tests (Ginkgo + framework vs. Go unit + ktesting) is something that can evolve in the future. For now, the suite reuses helpers from test/e2e/dra/utils and thus is implemented as a Ginkgo suite.
    
The new helper code for managing a kind cluster is already written so that it could be used both with Go test and Ginkgo. https://github.com/kubernetes/kubernetes/pull/122481 could make that a bit easier, but is not absolutely required.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
E2E framework: support setting REST config

A special test (for example, one which manages its own cluster) could almost
construct a Framework instance because most fields are exported. The
`clientConfig` field isn't because REST configs often need to be deep-copied to
avoid accidentally updating some shared copy, so for this special case a
SetClientConfig is needed.

pohly committed:
DRA integration: add upgrade/downgrade testing with local-up-cluster.sh

The test brings up the cluster and uses that power to run through
an upgrade/downgrade scenario. Version skew testing (running tests while the cluster
is partially up- or downgraded) could be added.

The new helper code for managing the cluster is written so that it could be
used both in an integration test and an E2E
test. https://github.com/kubernetes/kubernetes/pull/122481 could make that a
bit easier in an E2E test, but is not absolutely required.

In contrast to running on a normal cluster, pods need no privileges.
Instead, the caller has to make sure that the test itself can write
into system directories used by the cluster.

pohly committed:
ktesting: skip logging error when capturing it

Hiding the error in WithError is the right choice for example
when it is used inside ktesting.Eventually. Most callers probably want to deal
with the unexpected error themselves. For those who don't, WithErrorLogging
continues to log it.

pohly committed:
DRA E2E: increase verbosity of resource claim controller

Can be done via -vmodule, albeit not precisely because other controllers
also have a controller.go file.

pohly committed:
DRA E2E: reduce port proxy verbosity

During upgrade/downgrade testing, errors are encountered while the apiserver is
down. This is normal and handled via retrying, so we don't need to be verbose.

pohly committed:
E2E framework: retry also on EOF errors

EOF occurs after restarting the API server and, despite a retry loop in
client-go/rest/request.go, sometimes is returned to the application.

pohly committed:
ktesting: add step Begin/End

This allows declaring a code region as one step without having to use
an anonymous callback function, which has the advantage that variables
set during the step are visible afterwards.

In Python, this would be done as

    with ktesting.Step(tctx) as tcxt:
        // some code code inside step
    // code not in the same step

But Go has no such construct.

In contrast to WithStep, the start and end of the step are logged, including
timing information.

pohly committed:
DRA E2E: support using ktesting

This is a DRA-specific stop-gap solution for using the E2E framework together
with ktesting. Long-term this should better land in the E2E framework itself.

pohly commented:
/test pull-kubernetes-kind-dra

pohly commented on test/e2e/dra/versionskew.go:
```diff
@@ -0,0 +1,263 @@
+		tCtx = ktesting.End(tCtx)
+
+		clusterName := "dra"
+		tCtx = ktesting.Begin(tCtx, fmt.Sprintf("bring up v%d.%d", major, previousMinor))
```
I'm still undecided whether helper code like the new `kindcluster` should raise assertions itself or return an error that the caller must wrap to provide more context.

In this PR I am using "steps" to provide that context and avoid returning errors. Not having to handle those makes the test simpler and the step context also shows up in log entries, which helps when the same log message is emitted in different parts of the same test:

```
  I0617 17:34:42.433225 951392 stepcontext.go:75] bring up v1.33: Starting...
  I0617 17:34:42.876173 951392 versionskew.go:83] bring up v1.33: Running command: kind create cluster --retain --name dra --config /tmp/ginkgo3832759802/kind.yaml --kubeconfig /tmp/ginkgo3832759802/kube.config --image kind/cluster:1.33
  I0617 17:34:43.008304 951392 kindcluster.go:452] bring up v1.33: kind: Creating cluster "dra" ...
...
  I0617 17:35:08.992754 951392 stepcontext.go:86] bring up v1.33: Done, duration 26.5595254s
```

The downsides of steps is that it's easier to forget about setting them up, compared to the more usual error wrapping, and then assertion failures lack additional context.

pohly commented on test/e2e/feature/feature.go:
```diff
@@ -208,9 +208,18 @@ var (
+	// They must support bringing up a cluster with a node image built from
+	// the current Kubernetes version on the host on which the E2E suite runs.
+	// The repo root must point to the Kubernetes source code.
+	KindCommand = framework.WithFeature(framework.ValidFeatures.Add("KindCommand"))
```
@aojea: does this look reasonable to you?

From the commit message:

It is debatable whether this should be an E2E test at all. Technically this could also be an integration test. It's currently done mostly for pragmatic reasons:

- It can reuse the existing helper code.
- It can run in the normal DRA E2E jobs, with timeouts defined there.
- Integration tests are limited to 10 minutes per directory and
      pull-kubernetes-integration is already a big and slow job.
      A different approach for per-SIG or per-feature integration
      testing would be needed.

pohly replied:
Or simply leave it where it is and turn it into a Go unit test which skips itself unless a command line flag or env variable enables it...

I'm going to try that.

pohly replied:
New job: https://github.com/kubernetes/test-infra/pull/35004

aojea replied:
The e2e test framework, should focus on testing the application's functionality, and now becomes responsible for provisioning and de-provisioning the test environment. This blurs responsibilities violating the separation of concerns.

I really do not like this coupling

> If we don't build this into the existing e2e.test, then we will need a new Ginkgo or Go testsuite and some mechanism to invoke it, including separate jobs. I'm not sure whether it's worth it.

This already exist https://github.com/kubernetes/kubeadm/blob/main/kinder/doc/test-upgrades.md

kubeadm folks were testing upgrades for a long time https://testgrid.k8s.io/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-upgrade-1-33-latest

aojea replied:
> * In-tree, because then we can modify tests together with the thing they are testing. Feature graduation is likely to require updating the tests.

cluster lifecyle can not be part of the test framework, or you can deadlock

pohly replied:
This is already a problem that we have, which is why I asked on Slack the other day about how to handle changes in Kubernetes which break kind: kind is used in merge-blocking jobs, preventing merging such breaking changes. As explained on Slack, first kind has to be updated to support the upcoming change and the current Kubernetes, jobs use the latest kind, then the PR with the breaking change can be merged.

The usage of kind here is not different, the experimental job also [uses the latest kind](https://github.com/kubernetes/test-infra/blob/ec8a6bb8e952385b2becf91b1e02a04c104d10b8/config/jobs/kubernetes/sig-node/dra-canary.yaml#L416-L417).

The Go code in `test/utils/kindcluster` uses similar kind APIs as other jobs, with one exception: to replace components, it has to make assumptions about how control plane components are started (static pods), where the kubelet is located in the node image, and that there is a systemd unit for it.

This risk can be mitigated by not forking the job which runs the upgrade/downgrade tests for older releases, not making it merge-blocking and optional (also a good choice because of costs!) and not release-informing/blocking. Then if something really breaks, we can fix it in the master branch without affecting anyone.

As long as it doesn't break because of kind changes, we get a CI signal that we don't have right now ("are upgrades and downgrades involving DRA (or any other new feature) currently working?").

pohly replied:
If this usage of kind as an out-of-tree component is a blocker, than I can pivot to use `hack/local-up-cluster.sh` instead. That's in-tree and should also be usable. I was considering that initially, but then used kind and https://gist.github.com/aojea/2c94034f8e86d08842e5916231eb3fe1#file-kind_upgrade-sh-L27 because it's nicer and produces a multi-node cluster.

BenTheElder replied:
The main issue with local-up-cluster.sh is maintainers, @dims was the most active and has recently pointed people to kind. It's also a bit difficult for using locally, it runs locally and isn't sandboxed so it manipulates the host directly quite a bit, in CI of course we stick that inside a container and it's similar to kind ...

> The Go code in test/utils/kindcluster uses similar kind APIs as other jobs, with one exception: to replace components, it has to make assumptions about how control plane components are started (static pods), where the kubelet is located in the node image, and that there is a systemd unit for it.

I'm not happy with kubernetes/kubernetes code making assumptions about kind internals, if they ever change in kind then we need to patch it in every active release branch. This can totally deadlock both projects. We very explicitly tell users not to depend on details like "what is the path to kubelet", see here:
https://kind.sigs.k8s.io/docs/design/node-image/

> We only support that node images will create a working Kubernetes node at the advertised version with the kind version they were released with (and best effort with other releases), see the release notes.
> 
> The contents and implemlentation of the images are subject to change at any time to fix bugs, improve reliability, performance, or maintainability.
>
> DO NOT DEPEND ON THE INTERNALS OF THE NODE IMAGES.
>
> KIND provides [conformant](https://www.cncf.io/training/certification/software-conformance/) Kubernetes, anything else is an implementation detail.
>
> We will not accept bugs about “breaking changes” to node images and you depend on the implementation details at your own peril.

At least when test-infra experimental scripts leverage these, we can patch it once across release branches.

pohly commented on test/e2e/dra/versionskew.go:
```diff
@@ -0,0 +1,263 @@
+
+var _ = framework.SIGDescribe("node")(framework.WithLabel("DRA"), feature.DynamicResourceAllocation, framework.WithFeatureGate(features.DynamicResourceAllocation), framework.WithSlow() /* TODO: add feature.KindCommand once jobs are updated*/, func() {
+
+	ginkgo.It("upgrade/downgrade works", func(ctx context.Context) {
```
@lauralorenz, @SergeyKanzhelev: once this PR is merged with this initial base test, it would be great to extend it with other scenarios described in https://github.com/kubernetes/kubernetes/issues/128965.

pohly commented on test/e2e/dra/versionskew.go:
```diff
@@ -0,0 +1,263 @@
+// Beware that the apiserver not coming up again can be caused by not
+// having up-to-date release images, so run "make quick-release" manually.
+
+var _ = framework.SIGDescribe("node")(framework.WithLabel("DRA"), feature.DynamicResourceAllocation, framework.WithFeatureGate(features.DynamicResourceAllocation), framework.WithSlow() /* TODO: add feature.KindCommand once jobs are updated*/, func() {
```
@aojea: we need to agree on this new feature, then I need to whitelist it in the DRA jobs, then use it here.

We do not absolutely need it for DRA (tests currently only run in jobs which support kind), but it would be cleaner to have it.

pohly reviewed (commented):
Seeding the review by calling out some aspects...

pohly commented on pkg/controller/resourceclaim/controller.go:
```diff
@@ -338,7 +338,7 @@ func (ec *Controller) podNeedsWork(pod *v1.Pod) (bool, string) {
 
 func (ec *Controller) enqueueResourceClaim(logger klog.Logger, oldObj, newObj interface{}) {
-	deleted := newObj != nil
+	deleted := newObj == nil
```
My thoughts exactly :cry:

https://github.com/kubernetes/kubernetes/issues/132334

Included here as a stop-gap, not meant to be merged via this PR.

nojnhuh replied:
I opened #132533 with this fix.

pohly commented:
/test pull-kubernetes-kind-dra

Temporarily removed the `WithSlow` to get it to run...

pohly commented:
/test pull-kubernetes-kind-dra

pohly commented:
/test pull-kubernetes-kind-dra

pohly commented:
/test pull-kubernetes-kind-dra pull-kubernetes-unit pull-kubernetes-integration

pohly committed:
DRA E2E: retry exec of hostpathplugin

We can recover from exec failing, the portproxy code already retries port
forwarding.

pohly commented:
/test pull-kubernetes-dra-integration-canary

pohly commented on test/utils/kindcluster/kindcluster.go:
```diff
@@ -0,0 +1,486 @@
+// Cluster represents one kind cluster. Each instance is
+// independent from all other instances as long the name
+// chosen by the caller is unique on the node.
+type Cluster struct {
+	running    bool
+	name       string
+	dir        string
+	kubeConfig string
+}
+
+// Start brings up the cluster anew. If it was already running, it will be stopped first.
+
+// Will be stopped automatically at the end of the test.
+// If the ARTIFACTS env variable is set and the test failed,
+// log files of the kind cluster get dumped into
+// $ARTIFACTS/<test name>/kind/<cluster name> before stopping it.
+//
+// The name should be unique to enable running clusters in parallel.
+// The config has to be a complete YAML according to https://kind.sigs.k8s.io/docs/user/configuration/.
+// The image source can be a directory containing Kubernetes source code or
+// a download URL like https://dl.k8s.io/ci/v1.33.1-19+f900f017250646/kubernetes-server-linux-amd64.tar.gz.
+func (c *Cluster) Start(tCtx ktesting.TContext, name, kindConfig, imageName string) {
```
Longer discussion in https://github.com/kubernetes/kubernetes/pull/132295#discussion_r2166097335 (helpfully hidden by GitHub when loading the page, thanks GitHub! :angry: ).

BenTheElder replied:
Yeah ... I'm not convinced the e2e framework is the right place to maintain logic like this, which will likely span releases etc.

This is developing really tight coupling to kind internals.

BenTheElder commented on test/e2e/dra/versionskew_test.go:
```diff
@@ -0,0 +1,302 @@
+// Adapted from https://github.com/kubernetes-sigs/kind/blob/3df64e784cc0ea74125b2a2e9877817418afa3af/pkg/build/nodeimage/internal/kube/source.go#L71-L104
+func sourceVersion(tCtx ktesting.TContext, kubeRoot string) (gitVersion string, dockerTag string, err error) {
+	// Get the version output.
+	cmd := exec.CommandContext(tCtx, "hack/print-workspace-status.sh")
```
I don't think e2e.test should be doing this, or compiling anything. This is something we'd do in a tool like kubetest2 or kinder. It's a slow, orthogonal concern to testing.

BenTheElder reviewed (changes requested):
I don't think this is the right approach. We should keep the tests in e2e.test generic and put logic like upgrading clusters into external tools (like [kinder](https://github.com/kubernetes/kubeadm/blob/main/kinder/doc/kind-kinder.md), or https://github.com/kubernetes/test-infra/tree/master/experiment/compatibility-versions

I'm pretty concerned about having kubernetes/kubernetes take dependencies on internals we've repeatedly and explicitly told users not to use and documented as such. kind is tricky to maintain as-is.

pohly committed:
ktesting: WithCancel does not cancel on test completion anymore

That WithCancel added a deferred cleanup which cancels on test termination was
unexpected. This automatic cancellation makes sense only for the initial root
TContext.

pohly committed:
ktesting: document side effect of SetDefaultVerbosity during init

pohly committed:
ktesting: add WithoutCancel

This closes a gap compared to the context package. It's useful when combined
with Ginkgo to keep something running beyond the end of the Ginkgo BeforeEach
or It node.

pohly committed:
DRA E2E: support deploying driver on localhost

If we know that the test binary shares the filesystem with the cluster (for
example, when using local-up-cluster.sh), then we can avoid the whole
complicated portproxy solution and work directly with the paths on the
host.

Only works with suitable permissions! /var/lib/kubelet/plugins,
/var/lib/kubelet/plugin_registry, and /var/run/cdi must be writable.

portproxy remains the default because it automatically gains sufficient
permissions also when combined with local-up-cluster.sh.

pohly committed:
DRA E2E: refactor helper code

The helper code is useful for a separate Ginkgo suite for upgrade/downgrade
testing. We don't want to import test/e2e/dra there because that would also
define additional tests.

pohly commented:
/test pull-kubernetes-local-e2e

pohly commented:
/test pull-kubernetes-dra-integration-canary

pohly commented:
/test pull-kubernetes-dra-integration-canary

pohly commented:
/test pull-kubernetes-dra-integration-canary

pohly commented:
/test pull-kubernetes-dra-integration-canary

bart0sh commented:
/test pull-kubernetes-dra-integration-canary

pohly commented:
@bart0sh fixed the setup of Docker (or rather, containerd) in the CI over in https://github.com/kubernetes/kubernetes/pull/132734. I've included his commits here.

/test pull-kubernetes-local-e2e
/test pull-kubernetes-dra-integration-canary
/test pull-kubernetes-verify

pohly commented:
/test pull-kubernetes-local-e2e
/test pull-kubernetes-dra-integration-canary
/test pull-kubernetes-verify

pohly commented:
/test pull-kubernetes-dra-integration-canary

pohly commented:
/test pull-kubernetes-dra-integration-canary

pohly commented:
/test pull-kubernetes-dra-integration-canary
/test pull-kubernetes-verify
/test pull-kubernetes-linter-hints
/test pull-kubernetes-local-e2e

> Darn "kube-proxy not started" flake...

This was a race between printing the output and printing "etcd running" because the `run` sub-shell is spawned in the background. Adding `wait` should fix this.

aojea commented on test/integration/dra/cluster/README.md:
```diff
@@ -0,0 +1,43 @@
+binaries, relying on local-up-cluster.sh for the logic and configuration
+steps. Because local-up-cluster.sh needs additional permissions and
+preparations on the host, the test cannot run in "make test-integration" and
+just skips itself there.
```
why not put it outside of integration folder then?

pohly replied:
It being under `test/integration/dra` has the practical advantage that all DRA integration tests, including this one, can be run with one command when setting up the environment accordingly.

But more importantly, I couldn't think of any other directory that would fit better. `test/e2e` is worse. A new `test/dra`? `test/other-integration/dra`? It would need a better name. Either way, we then have to add exceptions for not running the test in `make test` or keeping the self-skip, so it wouldn't be substantially different from where it is now.

Any suggestion?

aojea replied:
It seems you have done the exercise already, maybe @BenTheElder has some other idea, I do not know if e2e_dra can be an option, I can see we have e2e_kubeadm and e2e_node already

pohly replied:
We could do `test/e2e_dra`. Setting it up would imply modifying `hack/make-rules/test.sh`, but that's a one-time thing. No strong opinion from me.

pohly replied:
The boundary between E2E and integration is indeed fluid. It's about testing end-to-end flows (E2E) with a combination of several different components (integration).

I can imagine that this may become useful also in other areas, but for now I prefer to focus on DRA and explore a bit more how to write such tests before opening it up for others.

pohly committed:
DRA E2E: move upgrade/downgrade test into test/e2e_dra

It's similar to test/e2e_kubeadm in the sense that it is a test which must be
excluded from both "make test" and "make integration" by default.

aojea commented:
/lgtm

pohly commented:
/hold

Let's test via an updated canary job before merging.

Also, failing the test instead of skipping it is better now that it doesn't run anymore in other jobs: https://github.com/kubernetes/kubernetes/compare/c475654a5dfa338c7f06ba6fecee77fe29dd0d95..56a004bf724f8012114ae668167dfaf9f202331f

pohly committed:
local-up-cluster.sh: don't require to be invoked in the root

It's normal for shell scripts to change the current directory if they expect to
run in the repo root.

pohly commented:
/test pull-kubernetes-dra-integration-canary

May have to rename the job...

pohly commented:
/test pull-kubernetes-dra-integration-canary

pohly commented:
Rebased because some new code in test/e2e/dra.go from master had to be updated.

/test pull-kubernetes-dra-integration-canary

aojea commented:
/lgtm

pohly commented:
/hold cancel

pacoxu commented:
/test pull-kubernetes-e2e-gce
